File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/stored/library/items/files/fragment/ActiveFileDownloadsFragment.java
package com.lasthopesoftware.bluewater.client.stored.library.items.files.fragment;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.content.LocalBroadcastManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.lasthopesoftware.bluewater.R;
import com.lasthopesoftware.bluewater.client.library.access.LibraryRepository;
import com.lasthopesoftware.bluewater.client.library.access.SelectedBrowserLibraryProvider;
import com.lasthopesoftware.bluewater.client.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.servers.selection.SelectedBrowserLibraryIdentifierProvider;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.StoredFileAccess;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.fragment.adapter.ActiveFileDownloadsAdapter;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.repository.StoredFile;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.retrieval.StoredFilesCollection;
import com.lasthopesoftware.bluewater.client.stored.worker.SyncWorker;
import com.lasthopesoftware.bluewater.shared.promises.extensions.LoopedInPromise;
import com.namehillsoftware.handoff.promises.response.VoidResponse;

import java.util.Map;

public class ActiveFileDownloadsFragment extends Fragment {

	private BroadcastReceiver onSyncStartedReceiver;
	private BroadcastReceiver onSyncStoppedReceiver;
    private BroadcastReceiver onFileQueuedReceiver;
    private BroadcastReceiver onFileDownloadedReceiver;
    private LocalBroadcastManager localBroadcastManager;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

	    localBroadcastManager = LocalBroadcastManager.getInstance(getActivity());
    }

	@Nullable
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		final RelativeLayout viewFileslayout = (RelativeLayout) inflater.inflate(R.layout.layout_downloads, container, false);

		final ProgressBar progressBar = viewFileslayout.findViewById(R.id.pbLoadingItems);
		final ListView listView = viewFileslayout.findViewById(R.id.lvItems);

		listView.setVisibility(View.INVISIBLE);
		progressBar.setVisibility(View.VISIBLE);

		final FragmentActivity activity = getActivity();
		if (activity == null) return viewFileslayout;

		final LibraryRepository libraryRepository = new LibraryRepository(activity);
		final SelectedBrowserLibraryProvider selectedBrowserLibraryProvider = new SelectedBrowserLibraryProvider(
			new SelectedBrowserLibraryIdentifierProvider(activity),
			libraryRepository);

		selectedBrowserLibraryProvider
			.getBrowserLibrary()
			.then(new VoidResponse<>(library -> {
				final StoredFileAccess storedFileAccess = new StoredFileAccess(
					activity,
					new StoredFilesCollection(activity));

				storedFileAccess.getDownloadingStoredFiles()
					.eventually(LoopedInPromise.response(new VoidResponse<>(storedFiles -> {
						final Map<Integer, StoredFile> localStoredFiles =
							Stream.of(storedFiles)
								.filter(f -> f.getLibraryId() == library.getId())
								.collect(Collectors.toMap(StoredFile::getId));

						final ActiveFileDownloadsAdapter activeFileDownloadsAdapter = new ActiveFileDownloadsAdapter(activity, localStoredFiles.values());

						if (onFileDownloadedReceiver != null)
							localBroadcastManager.unregisterReceiver(onFileDownloadedReceiver);

						onFileDownloadedReceiver = new BroadcastReceiver() {
							@Override
							public void onReceive(Context context, Intent intent) {
								final int storedFileId = intent.getIntExtra(SyncWorker.storedFileEventKey, -1);

								final StoredFile storedFile = localStoredFiles.get(storedFileId);
								if (storedFile == null) return;

								activeFileDownloadsAdapter.remove(new ServiceFile(storedFile.getServiceId()));
								localStoredFiles.remove(storedFileId);
							}
						};

						localBroadcastManager.registerReceiver(onFileDownloadedReceiver, new IntentFilter(SyncWorker.onFileDownloadedEvent));

						if (onFileQueuedReceiver != null)
							localBroadcastManager.unregisterReceiver(onFileQueuedReceiver);

						onFileQueuedReceiver = new BroadcastReceiver() {
							@Override
							public void onReceive(Context context, Intent intent) {
								final int storedFileId = intent.getIntExtra(SyncWorker.storedFileEventKey, -1);
								if (storedFileId == -1) return;

								if (localStoredFiles.containsKey(storedFileId)) return;

								storedFileAccess
									.getStoredFile(storedFileId)
									.eventually(LoopedInPromise.response(new VoidResponse<>(storedFile -> {
										if (storedFile == null || storedFile.getLibraryId() != library.getId())
											return;

										localStoredFiles.put(storedFileId, storedFile);
										activeFileDownloadsAdapter.add(new ServiceFile(storedFile.getServiceId()));
									}), activity));
							}
						};

						localBroadcastManager.registerReceiver(onFileQueuedReceiver, new IntentFilter(SyncWorker.onFileQueuedEvent));

						listView.setAdapter(activeFileDownloadsAdapter);

						progressBar.setVisibility(View.INVISIBLE);
						listView.setVisibility(View.VISIBLE);
					}), activity));
		}));

		final Button toggleSyncButton = viewFileslayout.findViewById(R.id.toggleSyncButton);
		final CharSequence startSyncLabel = activity.getText(R.string.start_sync_button);
		final CharSequence stopSyncLabel = activity.getText(R.string.stop_sync_button);

		SyncWorker.promiseIsSyncing()
			.eventually(LoopedInPromise.response(
				new VoidResponse<>(isSyncing -> toggleSyncButton.setText(!isSyncing ? startSyncLabel : stopSyncLabel)), activity));

		if (onSyncStartedReceiver != null)
			localBroadcastManager.unregisterReceiver(onSyncStartedReceiver);

		onSyncStartedReceiver = new BroadcastReceiver() {
			@Override
			public void onReceive(Context context, Intent intent) {
				toggleSyncButton.setText(stopSyncLabel);
			}
		};

		localBroadcastManager.registerReceiver(onSyncStartedReceiver, new IntentFilter(SyncWorker.onSyncStartEvent));

		if (onSyncStoppedReceiver != null)
			localBroadcastManager.unregisterReceiver(onSyncStoppedReceiver);

		onSyncStoppedReceiver = new BroadcastReceiver() {
			@Override
			public void onReceive(Context context, Intent intent) {
				toggleSyncButton.setText(startSyncLabel);
			}
		};

		localBroadcastManager.registerReceiver(onSyncStoppedReceiver, new IntentFilter(SyncWorker.onSyncStopEvent));

		toggleSyncButton.setOnClickListener(v -> SyncWorker.promiseIsSyncing()
			.then(isSyncing -> isSyncing
				? SyncWorker.cancel()
				: SyncWorker.syncImmediately(activity)));

		toggleSyncButton.setEnabled(true);

		return viewFileslayout;
	}

	@Override
    public void onDestroy() {
        super.onDestroy();

        if (localBroadcastManager == null) return;

		if (onSyncStartedReceiver != null)
			localBroadcastManager.unregisterReceiver(onSyncStartedReceiver);

		if (onSyncStoppedReceiver != null)
			localBroadcastManager.unregisterReceiver(onSyncStoppedReceiver);

        if (onFileDownloadedReceiver != null)
            localBroadcastManager.unregisterReceiver(onFileDownloadedReceiver);

        if (onFileQueuedReceiver != null)
            localBroadcastManager.unregisterReceiver(onFileQueuedReceiver);
    }
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/stored/library/items/files/fragment/ActiveFileDownloadsFragment.kt
package com.lasthopesoftware.bluewater.client.stored.library.items.files.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ProgressBar
import android.widget.RelativeLayout
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.lasthopesoftware.bluewater.R
import com.lasthopesoftware.bluewater.client.browsing.library.access.LibraryRepository
import com.lasthopesoftware.bluewater.client.browsing.library.access.session.SelectedBrowserLibraryIdentifierProvider
import com.lasthopesoftware.bluewater.client.browsing.library.access.session.SelectedBrowserLibraryProvider
import com.lasthopesoftware.bluewater.client.stored.library.items.files.StoredFileAccess
import com.lasthopesoftware.bluewater.client.stored.library.items.files.fragment.adapter.ActiveFileDownloadsAdapter
import com.lasthopesoftware.bluewater.client.stored.sync.StoredFileMessage
import com.lasthopesoftware.bluewater.client.stored.sync.SyncScheduler
import com.lasthopesoftware.bluewater.client.stored.sync.SyncStateMessage
import com.lasthopesoftware.bluewater.settings.repository.access.CachingApplicationSettingsRepository.Companion.getApplicationSettingsRepository
import com.lasthopesoftware.bluewater.shared.messages.application.ApplicationMessageBus.Companion.getApplicationMessageBus
import com.lasthopesoftware.bluewater.shared.messages.registerReceiver
import com.lasthopesoftware.bluewater.shared.promises.extensions.LoopedInPromise
import com.namehillsoftware.handoff.promises.Promise
import java.util.concurrent.ConcurrentHashMap

class ActiveFileDownloadsFragment : Fragment() {
	private var onSyncStartedReceiver: ((SyncStateMessage.SyncStarted) -> Unit)? = null
	private var onSyncStoppedReceiver: ((SyncStateMessage.SyncStopped) -> Unit)? = null
	private var onFileQueuedReceiver: ((StoredFileMessage.FileQueued) -> Unit)? = null
	private var onFileDownloadedReceiver: ((StoredFileMessage.FileDownloaded) -> Unit)? = null

	private val applicationMessageBus = lazy { requireContext().getApplicationMessageBus() }

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
		if (container == null) return null

		val viewFilesLayout = inflater.inflate(R.layout.layout_downloads, container, false) as RelativeLayout
		val progressBar = viewFilesLayout.findViewById<ProgressBar>(R.id.items_loading_progress)
		val listView = viewFilesLayout.findViewById<RecyclerView>(R.id.loaded_recycler_view)
		listView.visibility = View.INVISIBLE
		progressBar.visibility = View.VISIBLE

		val context = container.context
		val activeFileDownloadsAdapter = ActiveFileDownloadsAdapter(context)
		listView.adapter = activeFileDownloadsAdapter
		val layoutManager = LinearLayoutManager(context)
		listView.layoutManager = layoutManager

		listView.addItemDecoration(DividerItemDecoration(context, layoutManager.orientation))

		val applicationSettingsRepository = context.getApplicationSettingsRepository()
		val libraryRepository = LibraryRepository(context)
		val selectedBrowserLibraryProvider = SelectedBrowserLibraryProvider(
			SelectedBrowserLibraryIdentifierProvider(applicationSettingsRepository),
			libraryRepository)

		selectedBrowserLibraryProvider
			.browserLibrary
			.then { library ->
				val storedFileAccess = StoredFileAccess(
                    context
                )

				storedFileAccess.downloadingStoredFiles
					.eventually(LoopedInPromise.response({ storedFiles ->
						val localStoredFiles = ConcurrentHashMap(storedFiles.groupBy { sf -> sf.id }.values.associate { sf -> Pair(sf.first().id, sf.first()) })

						activeFileDownloadsAdapter.updateListEventually(localStoredFiles.values.toList())

						onFileDownloadedReceiver?.apply(applicationMessageBus.value::unregisterReceiver)
						applicationMessageBus.value
							.registerReceiver({ message: StoredFileMessage.FileDownloaded ->
								localStoredFiles.remove(message.storedFileId)
								activeFileDownloadsAdapter.updateListEventually(localStoredFiles.values.toList())
								Unit
							}.also { onFileDownloadedReceiver = it })

						onFileQueuedReceiver?.also(applicationMessageBus.value::unregisterReceiver)
						applicationMessageBus.value.registerReceiver({ message: StoredFileMessage.FileQueued ->
							message.storedFileId
								.takeUnless(localStoredFiles::containsKey)
								?.also { storedFileId ->
									storedFileAccess
										.getStoredFile(storedFileId)
										.eventually { storedFile ->
											if (storedFile != null && storedFile.libraryId == library?.id) {
												localStoredFiles[storedFileId] = storedFile
												activeFileDownloadsAdapter.updateListEventually(localStoredFiles.values.toList())
											} else {
												Promise.empty()
											}
										}
								}

							Unit
						}.also { onFileQueuedReceiver = it })

						progressBar.visibility = View.INVISIBLE
						listView.visibility = View.VISIBLE
					}, context))
			}

		val toggleSyncButton = viewFilesLayout.findViewById<Button>(R.id.toggleSyncButton)
		val startSyncLabel = context.getText(R.string.start_sync_button)
		val stopSyncLabel = context.getText(R.string.stop_sync_button)
		toggleSyncButton.isEnabled = false
		SyncScheduler.promiseIsSyncing(context).eventually(LoopedInPromise.response({ isRunning ->
			toggleSyncButton.text = if (!isRunning) startSyncLabel else stopSyncLabel
			toggleSyncButton.isEnabled = true
		}, context))

		onSyncStartedReceiver?.also(applicationMessageBus.value::unregisterReceiver)
		applicationMessageBus.value
			.registerReceiver({ _ : SyncStateMessage.SyncStarted -> toggleSyncButton.text = stopSyncLabel }.also { onSyncStartedReceiver = it })

		onSyncStoppedReceiver?.also(applicationMessageBus.value::unregisterReceiver)
		applicationMessageBus.value
			.registerReceiver({ _ : SyncStateMessage.SyncStopped -> toggleSyncButton.text = startSyncLabel }.also { onSyncStoppedReceiver = it })

		toggleSyncButton.setOnClickListener { v ->
			SyncScheduler.promiseIsSyncing(v.context).then { isSyncRunning ->
				if (isSyncRunning) SyncScheduler.cancelSync(v.context)
				else SyncScheduler.syncImmediately(context)
			}
		}
		return viewFilesLayout
	}

	override fun onDestroy() {
		super.onDestroy()

		if (applicationMessageBus.isInitialized()) {
			with (applicationMessageBus.value) {
				onSyncStartedReceiver?.also(::unregisterReceiver)
				onSyncStoppedReceiver?.also(::unregisterReceiver)
				onFileDownloadedReceiver?.also(::unregisterReceiver)
				onFileQueuedReceiver?.also(::unregisterReceiver)
			}
		}
	}
}
